### PR TITLE
searcher: add DB connection

### PR DESCRIPTION
**Pull request by sashaostrikov**
This is adding a DB connection to searcher. Connection is unused at the moment, but will be used when new features are implemented in scope of https://github.com/sourcegraph/sourcegraph/issues/27916

Closes https://github.com/sourcegraph/sourcegraph/issues/28260

### CODE REVIEW

File: cmd/searcher/main.go [hunk 1]
```diff
@@ -4,6 +4,9 @@ package main
 
 import (
 	"context"
+	"github.com/sourcegraph/sourcegraph/internal/conf/conftypes"
```
Comment (ryanslade) [inline cmd/searcher/main.go:7]: These should be grouped lower down with all our code

File: cmd/searcher/main.go [hunk 2]
```diff
@@ -67,6 +70,17 @@ func main() {
...
 
+	// DB initialization. It is unused at the moment, but will be used shortly
+	dsn := conf.GetServiceConnectionValueAndRestartOnChange(func(serviceConnections conftypes.ServiceConnections) string {
+		return serviceConnections.PostgresDSN
+	})
+	sqlDb, err := connections.NewFrontendDB(dsn, "searcher", false, &observation.TestContext)
+	if err != nil {
+		log.Fatalf("failed to initialize database store: %v", err)
+	} else {
+		log.Printf("Successfully initialized DB: %v", sqlDb)
```
Comment (ryanslade) [inline cmd/searcher/main.go:81]: I assume this is just here so that we use the variable?

File: cmd/searcher/main.go [hunk 3]
```diff
@@ -67,6 +70,17 @@ func main() {
 		cacheSizeBytes = i * 1000 * 1000
 	}
 
+	// DB initialization. It is unused at the moment, but will be used shortly
+	dsn := conf.GetServiceConnectionValueAndRestartOnChange(func(serviceConnections conftypes.ServiceConnections) string {
+		return serviceConnections.PostgresDSN
+	})
+	sqlDb, err := connections.NewFrontendDB(dsn, "searcher", false, &observation.TestContext)
```
Comment (ryanslade) [inline cmd/searcher/main.go:77]: nit: We usually just call this `db`

**Review by ryanslade — APPROVED** [review]
LGTM

Have you double checked that the dashboard appears as expected in your dev environment?

File: cmd/searcher/main.go (hunk 1 above)
Comment (sashaostrikov) [reply cmd/searcher/main.go:7]: Yes, I keep forgetting that Goland's "optimise imports" is not the same as `goimports` and I have to run latter manually

File: cmd/searcher/main.go (hunk 2 above)
Comment (sashaostrikov) [reply cmd/searcher/main.go:81]: exactly, the best "fake" usage I could imagine

**Comment by ryanslade** [comment]
Looks like some of our database connection code is changing soon, so should be aware of this: https://github.com/sourcegraph/sourcegraph/pull/29144

**Comment by sashaostrikov** [comment]
> Looks like some of our database connection code is changing soon, so should be aware of this: #29144

Thanks for noticing, I've added a TODO which I will check from time to time, to use new DB creation functions. For now using them is unsafe, I should wait for https://github.com/sourcegraph/sourcegraph/pull/29144 to be merged

**Comment by sashaostrikov** [comment]
PTAL
I've checked the dashboard locally -- it exists. PR is ready for re-review

File: cmd/searcher/main.go [hunk 4]
```diff
@@ -15,6 +15,10 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/sourcegraph/sourcegraph/internal/conf/conftypes"
```
Comment (ryanslade) [inline cmd/searcher/main.go:18]: Just the imports again 🤦 

I'm going to take a look into this now 😄

**Review by ryanslade — APPROVED** [review]
LGTM!

:shipit:

File: cmd/searcher/main.go (hunk 4 above)
Comment (sashaostrikov) [reply cmd/searcher/main.go:18]: it seems in the right place now, isn't it?

Comment (ryanslade) [reply cmd/searcher/main.go:18]: All the sourcegraph imports should be grouped together, so the `searcher` import below should also be there. 

We should have three groups:

* standard library
* other code
* our code

File: doc/admin/observability/dashboards.md [hunk 5]
```diff
@@ -10472,6 +10472,160 @@ Query: `sum by(instance) (rate(searcher_service_request_total[10m]))`
...
+
+#### searcher: max_open_conns
+
+<p class="subtitle">Maximum open</p>
+
+This panel has no related alerts.
+
+To see this panel, visit `/-/debug/grafana/d/searcher/searcher?viewPanel=100100` on your Sourcegraph instance.
+
+<sub>*Managed by the [Sourcegraph Core application team](https://handbook.sourcegraph.com/engineering/core-application).*</sub>
```
Comment (jplahn) [inline doc/admin/observability/dashboards.md:10485]: Core App team doesn't exist anymore. Are we going to change this and the below instances to Search Core (or Repo Management)?

Comment (sashaostrikov) [reply doc/admin/observability/dashboards.md:10485]: that's a nice point. I guess it should be Search Core but I'd double-check it with @ryanslade

Comment (ryanslade) [reply doc/admin/observability/dashboards.md:10485]: Yes, they should be owned by the search core team I believe. Worth letting them know in slack too.

Comment (sashaostrikov) [reply doc/admin/observability/dashboards.md:10485]: Okay, I'll make changes here and send a message in their channel

File: cmd/searcher/main.go [hunk 6]
```diff
@@ -67,6 +70,18 @@ func main() {
...
+	// DB initialization. It is unused at the moment, but will be used shortly
+	dsn := conf.GetServiceConnectionValueAndRestartOnChange(func(serviceConnections conftypes.ServiceConnections) string {
+		return serviceConnections.PostgresDSN
+	})
+	// TODO: use EnsureNewFrontendDB function as soon as https://github.com/sourcegraph/sourcegraph/pull/29144 is merged
+	db, err := connections.NewFrontendDB(dsn, "searcher", false, &observation.TestContext)
+	if err != nil {
+		log.Fatalf("failed to initialize database store: %v", err)
+	} else {
+		log.Printf("Successfully initialized DB: %v", db)
```
Comment (tsenart) [inline cmd/searcher/main.go:82]: nit: We don't need the else branch. For now you can just do `_, err := connections.NewFrontendDB(...)`

File: cmd/searcher/main.go [hunk 7]
```diff
@@ -67,6 +70,18 @@ func main() {
 		cacheSizeBytes = i * 1000 * 1000
 	}
 
+	// DB initialization. It is unused at the moment, but will be used shortly
+	dsn := conf.GetServiceConnectionValueAndRestartOnChange(func(serviceConnections conftypes.ServiceConnections) string {
+		return serviceConnections.PostgresDSN
+	})
+	// TODO: use EnsureNewFrontendDB function as soon as https://github.com/sourcegraph/sourcegraph/pull/29144 is merged
+	db, err := connections.NewFrontendDB(dsn, "searcher", false, &observation.TestContext)
```
Comment (tsenart) [inline cmd/searcher/main.go:78]: What is a observation.TestContext? Is there a non test context we should be using instead for production?

**Review by tsenart — COMMENTED** [review]
@ryanslade: With the decision to apply sub-repo perms filtering in the streaming.Sender in the frontend, do we still need to add all of these DB connection pools to the different services?

If we do, why? And also, we'll need to monitor the impact on the production DB when we roll this out. Until we have a centralized connection pooler like pg bouncer, adding services that connect directly to the DB is fragile.

**Comment by ryanslade** [comment]
> @ryanslade: With the decision to apply sub-repo perms filtering in the streaming.Sender in the frontend, do we still need to add all of these DB connection pools to the different services?
> 
> If we do, why? And also, we'll need to monitor the impact on the production DB when we roll this out. Until we have a centralized connection pooler like pg bouncer, adding services that connect directly to the DB is fragile.

I think you're right. We did need to add sub-repo filtering to the symbols client and some investigation in the past showed that we used that from searcher but I've just doubled checked and I don't see that dependency anymore. 

@sashaostrikov I think we can safely close this for now. If we discover later that we need to add a connection here at least the work is already done.

**Comment by sashaostrikov** [comment]
@ryanslade okay, got it. Then closing this PR

**Pull request closed without merging**